File: objc/PdDispatcher.m
//
//  PdDispatcher.m
//  libpd
//
//  Copyright (c) 2011 Peter Brinkmann (peter.brinkmann@gmail.com)
//
//  For information on usage and redistribution, and for a DISCLAIMER OF ALL
//  WARRANTIES, see the file, "LICENSE.txt," in this distribution.
//
//  Updated 2013, 2018 Dan Wilcox <danomatika@gmail.com>
//

#import "PdDispatcher.h"

@implementation PdDispatcher

- (instancetype)init {
	self = [super init];
	if (self) {
		listenerMap = [[NSMutableDictionary alloc] init];
		subscriptions = [[NSMutableDictionary alloc] init];
	}
	return self;
}

- (void)dealloc {
	for (NSValue *handle in subscriptions.allValues) {
		void *ptr = handle.pointerValue;
		[PdBase unsubscribe:ptr];
	}
}

- (int)addListener:(NSObject<PdListener> *)listener forSource:(NSString *)symbol {
	NSMutableArray *listeners = listenerMap[symbol];
	if (!listeners) {
		void *ptr = [PdBase subscribe:symbol];
		if (!ptr) {
			return -1;
		}
		NSValue *handle = [NSValue valueWithPointer:ptr];
		subscriptions[symbol] = handle;
		listeners = [[NSMutableArray alloc] init];
		listenerMap[symbol] = listeners;
	}
	[listeners addObject:listener];
	return 0;
}

- (int)removeListener:(NSObject<PdListener> *)listener forSource:(NSString *)symbol {
	NSMutableArray *listeners = listenerMap[symbol];
	if (listeners) {
		[listeners removeObject:listener];
		if (listeners.count == 0) {
			NSValue *handle = subscriptions[symbol];
			void *ptr = handle.pointerValue;
			[PdBase unsubscribe:ptr];
			[subscriptions removeObjectForKey:symbol];
			[listenerMap removeObjectForKey:symbol];
		}
	}
  return 0;
}

- (void)removeAllListeners {
	[listenerMap removeAllObjects];
	NSEnumerator *enumerator = [subscriptions objectEnumerator];
	id object;
	while (object = [enumerator nextObject]) {
		NSValue *handle = object;
		void *ptr = handle.pointerValue;
		[PdBase unsubscribe:ptr];
	}
	[subscriptions removeAllObjects];
}

// Override this method in subclasses if you want different printing behavior.
- (void)receivePrint:(NSString *)message {
	NSLog(@"Pd: %@\n", message);
}

- (void)receiveBangFromSource:(NSString *)source {
	NSArray *listeners = listenerMap[source];
	for (NSObject<PdListener> *listener in listeners) {
		if ([listener respondsToSelector:@selector(receiveBangFromSource:)]) {
		[listener receiveBangFromSource:source];
		} else {
		NSLog(@"Unhandled bang from %@", source);
		}
	}
}

- (void)receiveFloat:(float)received fromSource:(NSString *)source {
	NSArray *listeners = listenerMap[source];
	for (NSObject<PdListener> *listener in listeners) {
		if ([listener respondsToSelector:@selector(receiveFloat:fromSource:)]) {
			[listener receiveFloat:received fromSource:source];
		} else {
			NSLog(@"Unhandled float from %@", source);
		}
	}
}

- (void)receiveSymbol:(NSString *)symbol fromSource:(NSString *)source {
	NSArray *listeners = listenerMap[source];
	for (NSObject<PdListener> *listener in listeners) {
		if ([listener respondsToSelector:@selector(receiveSymbol:fromSource:)]) {
			[listener receiveSymbol:symbol fromSource:source];
		} else {
			NSLog(@"Unhandled symbol from %@", source);
		}
	}
}

- (void)receiveList:(NSArray *)list fromSource:(NSString *)source {
	NSArray *listeners = listenerMap[source];
	for (NSObject<PdListener> *listener in listeners) {
		if ([listener respondsToSelector:@selector(receiveList:fromSource:)]) {
			[listener receiveList:list fromSource:source];
		} else {
			NSLog(@"Unhandled list from %@", source);
		}
	}
}

- (void) receiveMessage:(NSString *)message withArguments:(NSArray *)arguments fromSource:(NSString *)source {
	NSArray *listeners = listenerMap[source];
	for (NSObject<PdListener> *listener in listeners) {
		if ([listener respondsToSelector:@selector(receiveMessage:withArguments:fromSource:)]) {
			[listener receiveMessage:message withArguments:arguments fromSource:source];
		} else {
			NSLog(@"Unhandled typed message from %@", source);
		}
	}
}

@end


@implementation LoggingDispatcher

-(void)receiveBangFromSource:(NSString *)source {
	NSLog(@"Received bang from source %@", source);
	[super receiveBangFromSource:source];
}

-(void)receiveFloat:(float)received fromSource:(NSString *)source {
	NSLog(@"Received float %f from source %@", received, source);
	[super receiveFloat:received fromSource:source];
}

-(void)receiveSymbol:(NSString *)symbol fromSource:(NSString *)source {
	NSLog(@"Received symbol %@ from source %@", symbol, source);
	[super receiveSymbol:symbol fromSource:source];
}

-(void)receiveList:(NSArray *)list fromSource:(NSString *)source {
	NSLog(@"Received list %@ from source %@", list, source);
	[super receiveList:list fromSource:source];
}

-(void)receiveMessage:(NSString *)message withArguments:(NSArray *)arguments fromSource:(NSString *)source {
	NSLog(@"Received message %@ with arguments %@ from source %@", message, arguments, source);
	[super receiveMessage:message withArguments:arguments fromSource:source];
}

@end

File: java/org/puredata/core/utils/PdDispatcher.java
/**
 * 
 * For information on usage and redistribution, and for a DISCLAIMER OF ALL WARRANTIES, see the
 * file, "LICENSE.txt," in this distribution.
 * 
 */

package org.puredata.core.utils;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.puredata.core.PdBase;
import org.puredata.core.PdListener;
import org.puredata.core.PdReceiver;

/**
 * 
 * PdDispatcher is an implementation of {@link PdReceiver} that dispatches messages from Pd to
 * instances of {@link PdListener} based on the Pd symbol they originate from. Instances of this
 * class automatically handle subscriptions to Pd symbols.
 * 
 * @author Peter Brinkmann (peter.brinkmann@gmail.com)
 * 
 */
public abstract class PdDispatcher implements PdReceiver {

  private final Map<String, Set<PdListener>> listeners = new HashMap<String, Set<PdListener>>();

  /**
   * Register a listener for Pd messages to the given symbol; subscribe to the symbol if this is the
   * first listener for this symbol
   * 
   * @param symbol receiver symbol in Pd
   * @param listener
   */
  public synchronized void addListener(String symbol, PdListener listener) {
    Set<PdListener> selected = listeners.get(symbol);
    if (selected == null) {
      int err = PdBase.subscribe(symbol);
      if (err != 0) {
        throw new IllegalArgumentException("bad symbol: " + symbol);
      }
      selected = new HashSet<PdListener>();
      listeners.put(symbol, selected);
    }
    selected.add(listener);
  }

  /**
   * Remove a listener associated with a symbol; unsubscribe from Pd messages to symbol if no
   * listeners are left
   * 
   * @param symbol receiver symbol in Pd
   * @param listener
   */
  public synchronized void removeListener(String symbol, PdListener listener) {
    Set<PdListener> selected = listeners.get(symbol);
    if (selected != null) {
      selected.remove(listener);
      if (selected.isEmpty()) {
        PdBase.unsubscribe(symbol);
        listeners.remove(symbol);
      }
    }
  }

  /**
   * Unsubscribe from Pd messages and release all resources
   */
  public synchronized void release() {
    for (String symbol : listeners.keySet()) {
      PdBase.unsubscribe(symbol);
    }
    listeners.clear();
  }

  @Override
  protected void finalize() throws Throwable {
    release();
    super.finalize();
  }

  /**
   * Handle print messages from Pd
   */
  @Override
  public abstract void print(String s);

  @Override
  public synchronized void receiveBang(String source) {
    Set<PdListener> selected = listeners.get(source);
    if (selected != null) {
      for (PdListener listener : selected) {
        listener.receiveBang(source);
      }
    }
  }

  @Override
  public synchronized void receiveFloat(String source, float x) {
    Set<PdListener> selected = listeners.get(source);
    if (selected != null) {
      for (PdListener listener : selected) {
        listener.receiveFloat(source, x);
      }
    }
  }

  @Override
  public synchronized void receiveSymbol(String source, String symbol) {
    Set<PdListener> selected = listeners.get(source);
    if (selected != null) {
      for (PdListener listener : selected) {
        listener.receiveSymbol(source, symbol);
      }
    }
  }

  @Override
  public synchronized void receiveList(String source, Object... args) {
    Set<PdListener> selected = listeners.get(source);
    if (selected != null) {
      for (PdListener listener : selected) {
        listener.receiveList(source, args);
      }
    }
  }

  @Override
  public synchronized void receiveMessage(String source, String symbol, Object... args) {
    Set<PdListener> selected = listeners.get(source);
    if (selected != null) {
      for (PdListener listener : selected) {
        listener.receiveMessage(source, symbol, args);
      }
    }
  }
}
